src/disableActEnvironment.ts

```diff
@@ -0,0 +1,171 @@
+
+const dispose: typeof Symbol.dispose =
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  Symbol.dispose ?? Symbol.for('nodejs.dispose')
+
+export interface DisableActEnvironmentOptions {
+  /**
+   * If `true`, all modifications of values set by `disableActEnvironment`
+   * will be prevented until `cleanup` is called.
+   *
+   * @default true
+   */
+  preventModification?: boolean
+
+  /**
+   * If `true`, will change the configuration of the testing library to
+   * prevent auto-wrapping e.g. `userEvent` calls in `act`.
+   *
+   * @default true
+   */
+  adjustTestingLibConfig?: boolean
+}
+
+/**
+ * Helper to temporarily disable a React 18+ act environment.
+ *
+ * By default, this also adjusts the configuration of @testing-library/dom
+ * to prevent auto-wrapping of user events in `act`, as well as preventing
+ * all modifications of values set by this method until `cleanup` is called
+ * or the returned `Disposable` is disposed of.
+ *
+ * Both of these behaviors can be disabled with the option, of the defaults
+ * can be changed for all calls to this method by modifying
+ * `disableActEnvironment.defaultOptions`.
+ *
+ * This returns a disposable and can be used in combination with `using` to
+ * automatically restore the state from before this method call after your test.
+ *
+ * @example
+ * ```ts
+ * test("my test", () => {
+ *   using _disabledAct = disableActEnvironment();
+ *
+ *   // your test code here
+ *
+ *   // as soon as this scope is left, the environment will be cleaned up
+ * })
+ * ```
+ *
+ * If you can not use the explicit resouce management keyword `using`,
+ * you can also manually call `cleanup`:
+ *
+ * @example
+ * ```ts
+ * test("my test", () => {
+ *   const { cleanup } = disableActEnvironment();
+ *
+ *   try {
+ *     // your test code here
+ *   } finally {
+ *     cleanup();
+ *   }
+ * })
+ * ```
+ *
+ * For more context on what `act` is and why you shouldn't use it in renderStream tests,
+ * https://github.com/reactwg/react-18/discussions/102 is probably the best resource we have.
+ */
+export function disableActEnvironment({
+  preventModification = disableActEnvironment.defaultOptions
+    .preventModification,
+  adjustTestingLibConfig = disableActEnvironment.defaultOptions
+    .adjustTestingLibConfig,
+}: DisableActEnvironmentOptions = {}): {cleanup: () => void} & Disposable {
+  const typedGlobal = globalThis as any as {IS_REACT_ACT_ENVIRONMENT?: boolean}
+  const cleanupFns: Array<() => void> = []
+
+  // core functionality
+  {
+    const previous = typedGlobal.IS_REACT_ACT_ENVIRONMENT
+    cleanupFns.push(() => {
+      Object.defineProperty(typedGlobal, 'IS_REACT_ACT_ENVIRONMENT', {
+        value: previous,
+        writable: true,
+        configurable: true,
+      })
+    })
+    Object.defineProperty(
+      typedGlobal,
+      'IS_REACT_ACT_ENVIRONMENT',
+      getNewPropertyDescriptor(false, preventModification),
+    )
+  }
```

Comment:
```suggestion
  const previous = typedGlobal.IS_REACT_ACT_ENVIRONMENT
  cleanupFns.push(() => {
    Object.defineProperty(typedGlobal, 'IS_REACT_ACT_ENVIRONMENT', {
      value: previous,
      writable: true,
      configurable: true,
    })
  })
  Object.defineProperty(
    typedGlobal,
    'IS_REACT_ACT_ENVIRONMENT',
    getNewPropertyDescriptor(false, preventModification),
  )
```

Are these brackets needed?

Reply:
Not 100% needed, but they are intentional here to restrict the scope of `previous` and group the core functionality (it also makes it collapsible in the editor as a bonus, which is something I actively use).